File: JS/Tree/105.Construct Binary Tree from Preorder and Inorder Traversal.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {number[]} preorder
 * @param {number[]} inorder
 * @return {TreeNode}
 */
var buildTree = function(preorder, inorder) {  // Just need to use inorder to find the root node, then we can split the given two number list as left and right part. Recursively perform the action until there's no elements left in the number list.
    if (preorder.length) {
        const rootValue = preorder[0];
        const rootIndex = inorder.indexOf(rootValue);
        
        const leftTreeInOrder = inorder.slice(0, rootIndex);
        const rightTreeInOrder = inorder.slice(rootIndex + 1);
      
        const leftTreePreOrder = preorder.slice(1, leftTreeInOrder.length + 1);
        const rightTreePreOrder = preorder.slice(leftTreeInOrder.length + 1);
      
        const root = new TreeNode(rootValue);
        root.left = buildTree(leftTreePreOrder, leftTreeInOrder);
        root.right = buildTree(rightTreePreOrder, rightTreeInOrder);
        return root;    
    }
    return null;
};

File: Golang/Tree/105.construct-binary-tree-from-preorder-and-inorder-traversal.go
package main

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func buildTree(preorder []int, inorder []int) *TreeNode {
	var indexOf = func(haystack []int, target int) int {
		for index, item := range haystack {
			if item == target {
				return index
			}
		}
		return -1
	}
	var recursive func(p, i []int) *TreeNode
	recursive = func(p, i []int) *TreeNode {
		var root *TreeNode
		length := len(p)
		if length > 0 {
			root = &TreeNode{}
			rootValue := p[0]
			root.Val = rootValue
			rootValueIndex := indexOf(i, rootValue)
			iLeft := i[:rootValueIndex]
			iRight := i[rootValueIndex+1:]
			pLeft := p[1 : len(iLeft)+1]
			pRight := p[len(iLeft)+1:]
			root.Left = buildTree(pLeft, iLeft)
			root.Right = buildTree(pRight, iRight)
		}
		return root
	}
	return recursive(preorder, inorder)
}
